src/app/server/Server.cpp

```diff
@@ -70,6 +72,89 @@ constexpr bool useTestPairing()
 #endif
 }
 
+CHIP_ERROR PersistAdminPairingToKVS(AdminPairingInfo * admin, AdminId nextAvailableId)
+{
+    ReturnErrorCodeIf(admin == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    ChipLogProgress(AppServer, "Persisting admin ID %d, next available %d", admin->GetAdminId(), nextAvailableId);
+
+    ReturnErrorOnFailure(admin->StoreUsingKVSMgr(PersistedStorage::KeyValueStoreMgr()));
+    ReturnErrorOnFailure(PersistedStorage::KeyValueStoreMgr().Put(kAdminTableCountKey, &nextAvailableId, sizeof(nextAvailableId)));
+
+    ChipLogProgress(AppServer, "Persisting admin ID successfully");
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR RestoreAllAdminPairingsFromKVS(AdminPairingTable & adminPairings, AdminId & nextAvailableId)
+{
+    // It's not an error if the key doesn't exist. Just return right away.
+    VerifyOrReturnError(PersistedStorage::KeyValueStoreMgr().Get(kAdminTableCountKey, &nextAvailableId) == CHIP_NO_ERROR,
+                        CHIP_NO_ERROR);
+    ChipLogProgress(AppServer, "Next available admin ID is %d", nextAvailableId);
+
+    for (AdminId id = 0; id < nextAvailableId; id++)
```

Comment:
I'm trying to understand AdminId.  It appears that there can be degenerate scenarios where many of the lower-numbered Ids are free, but `nextAvailableId` is close to `UINT16_MAX`, or perhaps no `nextAvailableId` is even available, despite the Id space being mostly empty.

I think persistence brings these degenerate scenarios closer to reality.  Such scenarios would also lead to excessive processing times scanning through many unused IDs in non-volatile storage.

Do we have any plans for a better system than just monotonically increasing AdminId?  It would be simple to solve these problems if we had some sort of relational database for our persistence.  That of course is off the table.  But even with the key=value store, there has to be a way to avoid these worst-case effects.

Reply:
Thinking about this a little more, one of the fundamental problems here is that the code in this PR must locate all entries in the key=value store with a given prefix.  But nothing in the storage delegate interface provides such a query.  In the absence of this, this code resorts to scanning all possible keys in the entire key space.  In many cases, this may be even worse than scanning all existing entries.

Since we obviously need it, why don't we add an interface to the storage delegate to execute operations against given key prefix wildcards?  This could be implemented trivially in the storage delegate base class by simply searching all entries for matches.  For platforms where optimization is possible, these methods can be overridden.

Reply:
`AdminId` allocation is done when a new ecosystem pairs with the device. Having a device cycle through `UINT16_MAX` admins without a factory reset may not be a practical scenario. I agree that it's doable, specially in test settings.

I don't think we have concluded how the `AdminId` space will be managed. I'll file a TODO to get back to it.

Reply:
@rgoliver could this be added as part of the KVS Manager?
`add an interface to the storage delegate to execute operations against given key prefix wildcards`
Reference this comment: https://github.com/project-chip/connectedhomeip/pull/4952/files#r579696949

Reply:
If we haven't decided how `AdminId` works, why are we persisting it to disk?

Reply:
This enables us to make progress in unrelated areas. e.g. if we want to reboot the device and still be able to use it from multiple admins, it'll enable that scenario. TODOs (which translate into GitHub issues) help keep track of the pending work. The pending work here is to decide how to allocate AdminId space. Rest of the AdminId APIs are functional.

Reply:
OK, makes sense.

I am merely pointing out that persisting things to disk for which the underlying design is incomplete is not free.  This has a definite likelihood of disruption if engineers are forced to factory reset and wipe devices when updating firmware.

But if this is what we have to do to hit milestones, I accept that.